### PR TITLE
ENH: Allow poorly formatted stata files to be read

### LINKED ISSUE #25960: UnicodeDecodeError for Stata file
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
pd.read_stata('mwe.dta')
```

`mwe.dta` available here: [mwe.zip](https://github.com/pandas-dev/pandas/files/3034141/mwe.zip)
This file is a derivative of [The Supreme Court Database](http://scdb.wustl.edu/_brickFiles/2018_02/SCDB_2018_02_caseCentered_Citation.dta.zip)

#### Problem description
The command raises

```python
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa4 in position 20: invalid start byte
```

I traced the error to a value label containing that byte.
This is a follow-up for #21244 and #23736
Changing line **1334** of `pandas.io.stata` from

```python
return s.decode('utf-8')
```

to

```python
return s.decode('latin-1')
```

allows me to read in the file.

#### Expected Output

The file should be correctly read and parsed.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.2.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 69 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.24.2
pytest: None
pip: 19.0.3
setuptools: 40.6.2
Cython: None
numpy: 1.16.2
scipy: None
pyarrow: None
xarray: None
IPython: 7.4.0
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by bashtage**
Add a fall back decode path that allows improperly formatted Stata
files written in 118 format but using latin-1 encoded strings to be
read

closes #25960

- [x] closes #25960
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/stata.py [hunk 1]
```diff
@@ -1327,12 +1327,20 @@ def _calcsize(self, fmt):
 
     def _decode(self, s):
         s = s.partition(b"\0")[0]
-        return s.decode('utf-8')
+        try:
+            return s.decode('utf-8')
```
Comment (WillAyd) [inline pandas/io/stata.py]: Out of curiosity why is this one hard coded where other methods seem to use `self._encoding`? Just trying to keep track of the various assignments and assumptions going on here

Comment (gfyoung) [reply pandas/io/stata.py]: Agreed, some commenting regarding the logic for this would be useful.

File: pandas/io/stata.py [hunk 2]
```diff
@@ -1327,12 +1327,20 @@ def _calcsize(self, fmt):
...
+        except UnicodeDecodeError:
+            # GH 25960
+            return s.decode('latin-1')
```
Comment (gfyoung) [inline pandas/io/stata.py]: Should we wrap this line in a `try-except` and provide a nicer error message regarding failure to decode?  Same question regarding the `try-except` you have for encoding.

Comment (bashtage) [reply pandas/io/stata.py]: It is rarely hit -- and only on an incorrectly formatted file, so I would say no.  There is nothing the user can realistically do to overcome a decode error now.

File: pandas/io/stata.py (hunk 1 above)
Comment (bashtage) [reply pandas/io/stata.py]: I refactored to remove null_terminate which isn't actually needed.

File: pandas/io/stata.py (hunk 2 above)
Comment (gfyoung) [reply pandas/io/stata.py]: Sure, but is the error message that you would get user-friendly?  The goal of the `try-except` that I am asking about is to produce a nicer error message, not necessarily to "overcome" the error.

Comment (bashtage) [reply pandas/io/stata.py]: I don't think this ever fails since 'latin-1' can decode all bytes.

Comment (bashtage) [reply pandas/io/stata.py]:
```
In [24]: b''.join([chr(i).encode('latin-1') for i in range(256)]).decode('latin1')

Out[24]: '\x00\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x0b\x0c\r\x0e\x0f\x10\x11\x12\x13\x14\x15
\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQR
STUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\x7f\x80\x81\x82\x83\x84\x85\x86\x87\x88\x89
\x8a\x8b\x8c\x8d\x8e\x8f\x90\x91\x92\x93\x94\x95\x96\x97\x98\x99\x9a\x9b\x9c\x9d\x9e\x9f
\xa0¡¢£¤¥¦§¨©ª«¬\xad®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïð
ñòóôõö÷øùúûüýþÿ'
```

Comment (bashtage) [reply pandas/io/stata.py]: I added a warning to let the user know that this is unexpected and to check the strings.

File: pandas/tests/io/test_stata.py [hunk 3]
```diff
@@ -1608,3 +1610,15 @@ def test_strl_latin1(self):
...
+
+    def test_encoding_latin1_118(self):
+        # GH 25960
+        msg = """
+One or more strings in the dta file could not be decoded using utf-8, and
+so the fallback encoding of latin-1 is being used.  This can happen when a file
+has been incorrectly encoded by Stata or some other software. You should verify
+the string values returned are correct."""
+        with pytest.warns(UnicodeWarning, match=msg):
+            encoded = read_stata(self.dta_encoding_118)
```
Comment (gfyoung) [inline pandas/tests/io/test_stata.py]: Use `tm.assert_produces_warning`.  Our implementation also checks stacklevel.

File: pandas/io/stata.py (hunk 2 above)
Comment (gfyoung) [reply pandas/io/stata.py]: That's fair.

**Comment by jreback** [comment]
lgtm. @bashtage if you can get the check_stacklevel to work would be great.

**Comment by bashtage** [comment]
I switched to `tm.assert_produces_warning`. Doesn't this automatically check the stacklevel?

https://github.com/pandas-dev/pandas/pull/25967/files#diff-ee04a1f9b23ca162f9592eef0caf0dbdR1621

File: pandas/tests/io/test_stata.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/io/test_stata.py]: actually this already checks

**Comment by jreback** [comment]
actually this does check, thanks @bashtage

**Pull request merged**